plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergPartitionEvolutionOnSameColumn.java

```diff
@@ -0,0 +1,169 @@
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class TestIcebergPartitionEvolutionOnSameColumn
+        extends AbstractTestQueryFramework
+{
+    private static final String BUCKET_NAME = "test-partition-evolution";
+
+    private Minio minio;
+
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        minio = closeAfterClass(Minio.builder().build());
+        minio.start();
+        minio.createBucket(BUCKET_NAME);
+
+        QueryRunner queryRunner = IcebergQueryRunner.builder()
+                .setIcebergProperties(
+                        ImmutableMap.<String, String>builder()
+                                .put("fs.native-s3.enabled", "true")
+                                .put("s3.aws-access-key", MINIO_ACCESS_KEY)
+                                .put("s3.aws-secret-key", MINIO_SECRET_KEY)
+                                .put("s3.region", MINIO_REGION)
+                                .put("s3.endpoint", minio.getMinioAddress())
+                                .put("s3.path-style-access", "true")
+                                .put("iceberg.register-table-procedure.enabled", "true")
+                                .buildOrThrow())
+                .build();
+
+        queryRunner.execute("CREATE SCHEMA IF NOT EXISTS " + ICEBERG_CATALOG + ".tpch");
+        return queryRunner;
+    }
+
+    @AfterAll
+    public void destroy()
+            throws Exception
+    {
+        minio = null; // closed by closeAfterClass
+    }
+
+    @Test
+    void testFilesPartitionEvolutionUsingTruncateOnSameColumn()
+    {
+        try (TestTable table = newTrinoTable("test_files_table", "(id int, part varchar) WITH (partitioning = ARRAY['truncate(part, 2)'])")) {
+            assertUpdate("INSERT INTO " + table.getName() + " VALUES (1, 'India')", 1);
+            assertThat(query("SELECT * FROM " + table.getName()))
+                    .matches("VALUES (1, VARCHAR 'India')");
+            assertThat(query("SELECT partition FROM \"" + table.getName() + "$files\""))
+                    .matches("VALUES ROW(CAST(ROW('In') AS ROW(part_trunc varchar)))");
+
+            assertUpdate("ALTER TABLE " + table.getName() + " SET PROPERTIES partitioning = ARRAY['truncate(part, 4)']");
+            assertUpdate("INSERT INTO " + table.getName() + " VALUES (2, 'Poland')", 1);
+            assertThat(query("SELECT * FROM " + table.getName()))
+                    .matches("VALUES (1, VARCHAR 'India'), (2, VARCHAR 'Poland')");
+            assertThat(query("SELECT partition FROM \"" + table.getName() + "$files\""))
+                    .matches("VALUES " +
+                            "ROW(CAST(ROW('In', null) AS ROW(part_trunc varchar, part_trunc_2 varchar))), " +
+                            "ROW(CAST(ROW(null, 'Pola') AS ROW(part_trunc varchar, part_trunc_2 varchar)))");
```

Comment:
Spark uses the bucket's number - `part_trunc_4` in this case. Is it possible to make it the same behavior?

Reply:
looking into it.